fix(upgrade): upgrade Directive facade should not return different instance from constructor

petebacondarwin opened the pull request:
In ngUpgrade (dynamic) we create a dynamic Angular `Directive` that wraps AngularJS components
that are being upgraded. The constructor of this `Directive` class returns a different instance
than `this`. It is this instance that actually contains the life-cycle hook handlers.

This would break in ivy, since the methods on the prototype of the original class are wired up,
rather than the instance methods. This results in hooks like `ngOnInit` not being called.

This PR refactors the code to extend the inner class that was being returned so that the
prototype chain is correct for both ViewEngine and ivy.

This change resolves a number of failing ivy tests, but also exposes other failures that were
masked by this issue. The tests have been updated accordingly.

petebacondarwin committed:
fix(upgrade): upgrade Directive facade should not return different instance from constructor

In ngUpgrade (dynamic) we create a dynamic Angular `Directive` that wraps AngularJS components
that are being upgraded. The constructor of this `Directive` class returns a different instance
than `this`. It is this instance that actually contains the life-cycle hook handlers.

This would break in ivy, since the methods on the prototype of the original class are wired up,
rather than the instance methods. This results in hooks like `ngOnInit` not being called.

This commit refactors the code to extend the inner class that was being returned so that the
prototype chain is correct for both ViewEngine and ivy.

This change resolves a number of failing ivy tests, but also exposes other failures that were
masked by this issue. The tests have been updated accordingly.

(FW-812)

petebacondarwin commented on packages/upgrade/test/common/test_helpers.ts:
```diff
@@ -9,14 +9,14 @@ import {setAngularJSGlobal} from '@angular/upgrade/src/common/angular1';
+  // {
+  //   label: '1.6',
+  //   files: ['angular-1.6/angular.js', 'angular-mocks-1.6/angular-mocks.js'],
+  // },
```
To be removed - this just stops the tests from being run 3 times each!

petebacondarwin commented on packages/upgrade/src/dynamic/upgrade_ng1_adapter.ts:
```diff
@@ -46,23 +46,13 @@ export class UpgradeNg1ComponentAdapterBuilder {
-    class MyClass {
-      // TODO(issue/24571): remove '!'.
-      directive !: angular.IDirective;
+    class MyClass extends UpgradeNg1ComponentAdapter implements OnInit, OnChanges, DoCheck,
```
It is probably overkill to add the `implements OnInint, OnChanges, DoCheck` clause but it felt good.

gkalpak replied:
Definitely not a hill I'm willing to die on :smiley:
(But hopefully, there are tests that would catch such breakages too :grin:)

petebacondarwin replied:
Actually (apart from `OnDestroy`) the `UpgradeNg1ComponentAdapter` already does have these interfaces attached.

petebacondarwin replied:
In terms of catching breakages, Angular doesn't actually require these interfaces to be attached at all. So we could just remove them altogether - and the tests will catch breakages :-P

mary-poppins commented:
You can preview 1ef6aa4 at https://pr27660-1ef6aa4.ngbuilds.io/.

mary-poppins commented:
You can preview c096bf7 at https://pr27660-c096bf7.ngbuilds.io/.

mary-poppins commented:
You can preview b809816 at https://pr27660-b809816.ngbuilds.io/.

mary-poppins commented:
You can preview 4a1e843 at https://pr27660-4a1e843.ngbuilds.io/.

gkalpak commented on packages/upgrade/src/dynamic/upgrade_ng1_adapter.ts:
```diff
@@ -46,23 +46,13 @@ export class UpgradeNg1ComponentAdapterBuilder {
+        OnDestroy {
       constructor(
           @Inject($SCOPE) scope: angular.IScope, injector: Injector, elementRef: ElementRef) {
-        const helper = new UpgradeHelper(injector, name, elementRef, this.directive);
```
Passing `this.directive` to `UpgradeHelper` was intentional (sort of - see below) as an optimization over having to retrieve the directive more than once (given that it is already retrieved in [UpgradeNg1ComponentAdapterBuilder.resolve()][1]).
The bug here is that we inteded to pass `UpgradeNg1ComponentAdapterBuilder`'s `.directive` and not `MyClass`' `.directive`.


I would rather fix that bug and not remove the optimization.

[1]: https://github.com/angular/angular/blob/4a1e843330021dae0f8cdc1e966696f940a3ecc3/packages/upgrade/src/dynamic/upgrade_ng1_adapter.ts#L126

gkalpak replied:
Apparently, I introduced the bug back in 11db3bd85 😇

gkalpak replied:
I don't think it does. `getDirective()` will still be called on `UpgradeNg1ComponentAdapterBuilds.resolve()` and in `UpgradeHelper` _every_ time we create a new instance of the upgraded component.

petebacondarwin replied:
Oh I see.... that is rather convoluted, eh?

petebacondarwin replied:
So I should just be able to stick `self.directive || undefined` as an extra parameter to `new UpgradeHelper` right?

gkalpak replied:
Right 👍

gkalpak reviewed (dismissed):
:tada:

petebacondarwin commented on packages/upgrade/test/static/integration/downgrade_component_spec.ts:
```diff
@@ -22,7 +22,7 @@ withEachNg1Version(() => {
     afterEach(() => destroyPlatform());
 
-    fixmeIvy('FW-716: Error: [$rootScope:inprog] $digest already in progress')
+    fixmeIvy('FW-715: ngOnChanges being called a second time unexpectedly')
```
This is just a side-update (sorry). It turns out that FW-716 is actual just another case of FW-715.

petebacondarwin commented on packages/upgrade/test/static/integration/upgrade_component_spec.ts:
```diff
@@ -3336,7 +3336,8 @@ withEachNg1Version(() => {
 
-      fixmeIvy('FW-713: ngDestroy not being called when downgraded ng2 component is destroyed')
+      fixmeIvy(
+          'FW-843: destroy hooks are not registered on upgraded ng1 components contained in ng2 component templates under ivy')
```
Again, a side update - just keeping the `fixmeIvy` descriptions in sync with the latest analysis

petebacondarwin committed:
test(upgrade): update test failure cause descriptions

The changes here clarify that we now have a better description
of the root cause of the failure.

mary-poppins commented:
You can preview 9cb16e0 at https://pr27660-9cb16e0.ngbuilds.io/.

petebacondarwin committed:
fixup! fix(upgrade): upgrade Directive facade should not return different instance from constructor

mary-poppins commented:
You can preview 0ba061f at https://pr27660-0ba061f.ngbuilds.io/.

mary-poppins commented:
You can preview 1f7c9bf at https://pr27660-1f7c9bf.ngbuilds.io/.

mhevery commented:
http://test/OCL:225893715:BASE:225893959:1545086831857:24be1aa6
